File: elixir/lib/medium/rectangles.ex
defmodule Rectangles do
  @doc """
  Count the number of ASCII rectangles.
  """
  @spec count(input :: String.t()) :: integer
  def count(input) do
    input
    |> parse_dots()
    |> find_rectangles()
    |> Enum.uniq()
    |> length()
  end

  defp parse_dots(input) do
    rows = String.split(input, ~r/\R/)

    rows
    |> Enum.with_index()
    |> Enum.flat_map(fn {col, y} ->
      col
      |> String.codepoints()
      |> Enum.with_index()
      |> Enum.map(fn {d, x} -> {d, x, y} end)
    end)
  end

  defp find_rectangles(dots) do
    corners = for {"+", x, y} <- dots, do: {x, y}

    Enum.flat_map(corners, &find_dot_rectangles(&1, corners, dots))
  end

  defp find_dot_rectangles({origin_x, origin_y}, corners, dots) do
    diagonals = for {x, y} <- corners, x > origin_x and y > origin_y, do: {x, y}

    diagonals_with_corners =
      Enum.filter(diagonals, fn {x, y} ->
        {origin_x, y} in corners and {x, origin_y} in corners
      end)

    diagonals_with_valid_connections =
      Enum.filter(diagonals_with_corners, fn {diag_x, diag_y} ->
        hor =
          Enum.filter(dots, fn {dot, x, y} ->
            dot in ["+", "-"] and (y == origin_y || y == diag_y) and (x > origin_x and x < diag_x)
          end)

        hor = hor |> length() == (diag_x - origin_x - 1) * 2

        ver =
          Enum.filter(dots, fn {dot, x, y} ->
            dot in ["+", "|"] and (x == origin_x || x == diag_x) and (y > origin_y and y < diag_y)
          end)

        ver = ver |> length() == (diag_y - origin_y - 1) * 2

        hor && ver
      end)

    Enum.map(diagonals_with_valid_connections, fn x -> {{origin_x, origin_y}, x} end)
  end
end

File: rust/src/hard/rectangles.rs
type Dot = (char, usize, usize);
type Coord = (usize, usize);
type Rectangle = (Coord, Coord);

pub fn count(lines: &[&str]) -> u32 {
    let dots = parse_dots(lines);

    let mut rectangles = find_rectangles(&dots);
    rectangles.sort_unstable();
    rectangles.dedup();

    rectangles.len() as u32
}

fn parse_dots(lines: &[&str]) -> Vec<Dot> {
    lines
        .iter()
        .enumerate()
        .flat_map(|(y, col)| col.char_indices().map(move |(x, ch)| (ch, x, y)))
        .collect()
}

pub fn find_rectangles(dots: &[Dot]) -> Vec<Rectangle> {
    let corners: Vec<Coord> = dots
        .iter()
        .filter_map(|&(ch, x, y)| if ch == '+' { Some((x, y)) } else { None })
        .collect();

    corners
        .iter()
        .flat_map(|&origin| find_dot_rectangles(origin, &corners, dots))
        .collect()
}

fn find_dot_rectangles(origin: Coord, corners: &[Coord], dots: &[Dot]) -> Vec<Rectangle> {
    let (origin_x, origin_y) = origin;

    let diagonals = corners
        .iter()
        .copied()
        .filter(|&(x, y)| x > origin_x && y > origin_y)
        .filter(|&(x, y)| corners.contains(&(origin_x, y)) && corners.contains(&(x, origin_y)));

    diagonals
        .filter(|&(diag_x, diag_y)| {
            let hor_count = dots
                .iter()
                .filter(|&&(ch, x, y)| {
                    (ch == '+' || ch == '-') &&
                    (y == origin_y || y == diag_y) &&
                    x > origin_x && x < diag_x
                })
                .count();

            let ver_count = dots
                .iter()
                .filter(|&&(ch, x, y)| {
                    (ch == '+' || ch == '|') &&
                    (x == origin_x || x == diag_x) &&
                    y > origin_y && y < diag_y
                })
                .count();

            hor_count == (diag_x - origin_x - 1) * 2 &&
            ver_count == (diag_y - origin_y - 1) * 2
        })
        .map(|diag| (origin, diag))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn one_rectangle() {
        #[rustfmt::skip]
        let input = &[
            "+-+",
            "| |",
            "+-+",
        ];

        let output = count(input);
        let expected = 1;
        assert_eq!(output, expected);
    }

    #[test]
    fn five_rectangles_with_shared_parts() {
        #[rustfmt::skip]
        let input = &[
            "  +-+",
            "  | |",
            "+-+-+",
            "| | |",
            "+-+-+",
        ];

        let output = count(input);
        let expected = 5;
        assert_eq!(output, expected);
    }
}


